File: src/rating_bot/rich_text.py
"""Plain text extraction from Bot API 10.1 rich messages.

A rich message carries no ``text``; its content lives in ``rich_message.blocks``,
where every text position is either a bare string, a styled run, or a custom
emoji standing in for a character. Flattening gives us something to measure the
length of, and something to log.
"""

from collections.abc import Iterator
from typing import Any

from aiogram.types import Message

# Keys whose string values are read by a human. Everything else in a block is
# structure: block types, file ids, dimensions, emoji ids.
_TEXT_KEYS = frozenset({"text", "caption", "alternative_text", "title"})


def _strings(node: Any, is_text: bool) -> Iterator[str]:
    if isinstance(node, str):
        if is_text:
            yield node
    elif isinstance(node, (list, tuple)):
        for item in node:
            yield from _strings(item, is_text)
    elif isinstance(node, dict):
        for key, value in node.items():
            yield from _strings(value, key in _TEXT_KEYS)


def rich_message_text(message: Message) -> str:
    """Every visible run of a rich message, one line per block."""
    rich = getattr(message, "rich_message", None)
    if rich is None:
        return ""
    lines = []
    for block in rich.blocks or ():
        line = "".join(_strings(block.model_dump(), is_text=False)).strip()
        if line:
            lines.append(line)
    return "\n".join(lines)


def message_text(message: Message) -> str:
    """The prose of a message, wherever Telegram chose to put it."""
    return message.text or message.caption or rich_message_text(message)

File: internal/tg/richtext.go
// Package tg is the bot itself: how an update is routed, what each command
// answers, and what the scheduled jobs send.
package tg

import (
	"encoding/json"
	"strings"

	"github.com/go-telegram/bot/models"
)

// textKeys are the keys whose string values a human reads. Everything else in
// a block is structure: block types, file ids, dimensions, emoji ids.
var textKeys = map[string]bool{
	"text": true, "caption": true, "alternative_text": true, "title": true,
}

// collectStrings walks any decoded JSON and gathers the prose out of it. A
// rich message carries no flat text, and every future block type is still a
// tree of the same four keys, so this needs no knowledge of block types.
func collectStrings(node any, isText bool, out *strings.Builder) {
	switch value := node.(type) {
	case string:
		if isText {
			out.WriteString(value)
		}
	case []any:
		for _, item := range value {
			collectStrings(item, isText, out)
		}
	case map[string]any:
		for key, item := range value {
			collectStrings(item, textKeys[key], out)
		}
	}
}

// flattenBlocks renders one line per block, which is enough to log a rich
// message and to measure how long it is.
func flattenBlocks(blocks []any) string {
	var lines []string
	for _, block := range blocks {
		var line strings.Builder
		collectStrings(block, false, &line)
		if trimmed := strings.TrimSpace(line.String()); trimmed != "" {
			lines = append(lines, trimmed)
		}
	}
	return strings.Join(lines, "\n")
}

// richMessageText flattens a rich message by round-tripping it through JSON,
// so the walk above sees exactly what Telegram sent.
func richMessageText(rich *models.RichMessage) string {
	if rich == nil {
		return ""
	}
	encoded, err := json.Marshal(rich.Blocks)
	if err != nil {
		return ""
	}
	var blocks []any
	if err := json.Unmarshal(encoded, &blocks); err != nil {
		return ""
	}
	return flattenBlocks(blocks)
}

// MessageText is the prose of a message, wherever Telegram chose to put it.
func MessageText(message *models.Message) string {
	if message == nil {
		return ""
	}
	if message.Text != "" {
		return message.Text
	}
	if message.Caption != "" {
		return message.Caption
	}
	return richMessageText(message.RichMessage)
}
